[pkg/ottl] Use generics to avoid context cast in getters and funcs

bogdandrutu opened the pull request:
This PR also improves number of allocations (by 2, if I am not wrong is because of the removing of the ottl.TransformContext interface use, which requires a heap allocation) and overall performance of the transform processor.

**Before:**

```shell
goos: darwin
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/processor/transformprocessor/internal/traces
cpu: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
BenchmarkTwoSpans
BenchmarkTwoSpans/no_processing
BenchmarkTwoSpans/no_processing-16         	  783144	      1394 ns/op	    1528 B/op	      31 allocs/op
BenchmarkTwoSpans/set_attribute
BenchmarkTwoSpans/set_attribute-16         	  713299	      1691 ns/op	    1832 B/op	      35 allocs/op
BenchmarkTwoSpans/keep_keys_attribute
BenchmarkTwoSpans/keep_keys_attribute-16   	  789517	      1626 ns/op	    1560 B/op	      33 allocs/op
BenchmarkTwoSpans/no_match
BenchmarkTwoSpans/no_match-16              	  799210	      1520 ns/op	    1560 B/op	      33 allocs/op
BenchmarkTwoSpans/inner_field
BenchmarkTwoSpans/inner_field-16           	  737241	      1566 ns/op	    1560 B/op	      33 allocs/op
BenchmarkTwoSpans/inner_field_both_spans
BenchmarkTwoSpans/inner_field_both_spans-16         	  763725	      1622 ns/op	    1592 B/op	      35 allocs/op
```

**After:**

```shell
goos: darwin
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/processor/transformprocessor/internal/traces
cpu: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
BenchmarkTwoSpans
BenchmarkTwoSpans/no_processing
BenchmarkTwoSpans/no_processing-16         	  868864	      1296 ns/op	    1480 B/op	      29 allocs/op
BenchmarkTwoSpans/set_attribute
BenchmarkTwoSpans/set_attribute-16         	  738514	      1609 ns/op	    1784 B/op	      33 allocs/op
BenchmarkTwoSpans/keep_keys_attribute
BenchmarkTwoSpans/keep_keys_attribute-16   	  800833	      1450 ns/op	    1512 B/op	      31 allocs/op
BenchmarkTwoSpans/no_match
BenchmarkTwoSpans/no_match-16              	  835160	      1394 ns/op	    1512 B/op	      31 allocs/op
BenchmarkTwoSpans/inner_field
BenchmarkTwoSpans/inner_field-16           	  796388	      1445 ns/op	    1512 B/op	      31 allocs/op
BenchmarkTwoSpans/inner_field_both_spans
BenchmarkTwoSpans/inner_field_both_spans-16         	  798840	      1539 ns/op	    1544 B/op	      33 allocs/op
```

TylerHelmuth reviewed (commented):
I'll do an in-depth review over the next week, but no feedback so far.

TylerHelmuth commented on pkg/ottl/expression.go:
```diff
@@ -18,85 +18,102 @@ import (
+	GetSpan() ptrace.Span
+}
+
+type MetricContext interface {
```
We also need a DataPointContext so that we can get rid of the `GetItem()` calls and casts in the datapoints context.

bogdandrutu replied:
Good catch to actually suggest me to remove them :))

For the `DataPointContext` part and the `GetItem` there, still cannot do it since it needs to be able to return 4 types of DataPoints. I will look into that later.

bogdandrutu replied:
I removed all of them, since we use the struct directly (I don't need any cast). In the resource/scope I have that interface, but the implementation is already in internal, so no need for this to be public.

TylerHelmuth commented on processor/transformprocessor/config.go:
```diff
@@ -40,35 +40,35 @@ var _ config.Processor = (*Config)(nil)
 	var errors error
 
-	ottlp := ottl.NewParser(
+	ottltrcesp := ottl.NewParser[ottltraces.TransformContext](
```
```suggestion
	ottltracesp := ottl.NewParser[ottltraces.TransformContext](
```

bogdandrutu replied:
Done.

bogdandrutu replied:
Done one more time :))

TylerHelmuth commented on processor/transformprocessor/config.go:
```diff
@@ -40,35 +40,35 @@ var _ config.Processor = (*Config)(nil)
 	}
 
-	ottlp = ottl.NewParser(
+	ottllogsp := ottl.NewParser[ottllogs.TransformContext](
```
Maybe a followup change, but instead of a generic NewParser, should contexts provide a Parser specific to their context?  Something like `ottllogs.NewParser(logs.Functions(), component.TelemetrySettings{}`)?

bogdandrutu replied:
I like that more, since 2 arguments are always coming from that package. Will address it in the next PR.

TylerHelmuth replied:
Sweet, do you plan to work on that or do we need an issue?

bogdandrutu replied:
Will do after you approve this 🗡️

TylerHelmuth commented on processor/transformprocessor/internal/logs/processor.go:
```diff
@@ -26,7 +26,7 @@ import (
 
 type Processor struct {
-	queries []ottl.Statement
+	queries []ottl.Statement[ottllogs.TransformContext]
```
Maybe we could do something similar here as proposed for `NewParser`?  Have `[]ottllogs.Statements` so that we don't have to have the processor pass in the Generic itself.

Hopefully we can make it so that the only place a processor would need to directly interact with the generic is when creating its own functions.

bogdandrutu replied:
So essentially have a `TracesStatements`?

bogdandrutu replied:
I can also look into this for the next PR when I look into the Parser.

TylerHelmuth replied:
Ya something like that.  Since the generic introduce some complexity, will be great to hide that complexity inside the Contexts as much as possible since, at least at the moment, we need a Parser and set of Statements per context.

TylerHelmuth commented:
@bogdandrutu I'll give this a final look over tomorrow.

bogdandrutu commented on pkg/ottl/contexts/ottldatapoints/datapoints.go:
```diff
@@ -26,6 +26,9 @@ import (
+var _ ottlcommon.ResourceContext = TransformContext{}
+var _ ottlcommon.InstrumentationScopeContext = TransformContext{}
+
 type TransformContext struct {
```
Done.

mcdoker18 committed:
[processor/transform] Added Int function to the transform processor (#14338)

[processor/transform] Added the Int function to the transform processor (#11810)

TylerHelmuth commented on exporter/awsxrayexporter/internal/translator/segment_test.go:
```diff
@@ -53,7 +53,7 @@ func TestClientSpanWithRpcAwsSdkClientAttributes(t *testing.T) {
 	attributes[conventions.AttributeHTTPTarget] = "/"
 	attributes[conventions.AttributeRPCService] = "DynamoDB"
-	attributes[conventions.AttributeRPCMethod] = "GetItem"
+	attributes[conventions.AttributeRPCMethod] = "GetDataPoint"
```
@bogdandrutu too ambitious of a replace haha

bogdandrutu replied:
:))))

evan-bradley reviewed (approved):
Looks good to me. Nice use of generics!
